accepted/future-releases/extension-types/feature-specification.md

```diff
@@ -1461,3 +1461,319 @@ wrapper class is so huge that it is likely to be a major use case for
+`dynamic` or a type which implements `Iterable<S>` for some `S`. That type may
+just also be an extension type._
+
+If `T` implements `Iterable<S>`, let `E` be `S`, otherwise let `E` be `Never` if
+`T` is a bottom type, or `dynamic` if `T` is `dynamic`.
+
+Now rewrite the `for`-`in` loop to:
+
+```dart
+Iterable<E> _$id1 = e;
+Iterator<E> _$id2 = _$id1.iterator;
+while (_$id2.moveNext()) {
+  D x = _$id2.current;
+  {
+    body
+  }  
+}
+```
+
+Like before, any errors that would occur in that program are reported as errors
+in the original `for`-`in` statement.
+
+_**TODO**: Specify this without using a “desugaring” rewrite. Problems like
+these are exactly why rewrites are problematic, they tend to introduce *more*
+information, through the extra syntax or inferred types, than what the original
+program warranted. Also, the rewrite supposedly happens after type inference, so
+we need to either perform type inference on the desugared code, or assign a
+context type and static type to each new expression, because those are
+referenced by the dynamic semantics._ 
+
+##### Consequences to existing code
+
+The only real change is that the type of `_$id1` is not the static type of `e`.
+That means that if there is a class which overrides its iterator to be more
+specific than required, that information is lost.
+
+Example:
+
+```dart
+class VeryInts extends Iterable<num> {
+  Iterator<int> get iterator => <int>[1].iterator;
+}
+void main() {
+  for (int i in VeryInts()) print(i);
+}
+```
+
+With the previous *specification*, the type that `.iterator` should be accessed
+on was `VeryInts`, which means that `_$id2` will get static type
+`Iterator<int>`, and its `.current` is then assignable to `int i`. With the new
+specification, we don’t look at the members of the actual type, instead we
+immediately up-cast to `Iterable<E>` and work with that.
+
+*Luckily*, our implementations already do that: CFE (Dart2js on DartPad):
+
+> ```
+> lib/main.dart:5:12:
+> Error: A value of type 'num' can't be assigned to a variable of type 'int'.
+> for (int i in VeryInts()) print(i);
+>         ^
+> ```
+
+and Analyzer:
+
+> ```
+> error: line 5 • The type 'VeryInts' used in the 'for' loop must implement 'Iterable' with a type argument that can be assigned to 'int'.
+> ```
+
+So, this looks like it might be a spec-only change for the static semantics, to
+make it match the actual implementation, and it’s *likely* that the
+implementations will also do the right thing when faced with extension types.
+They *must* only invoke instance `iterator`, `moveNext` and `current` members.
+
+#### Iterables and synchronous `yield*`
+
+The synchronous `yield*` operator takes an expression which, like `for`/`in`,
+must be assignable to `Iterable<T>`, and it too must use only instance members to
+iterate the operand where it state that it invokes `iterator`, `moveNext` and
+`current`.
+
+#### Stream and `await for`, asynchronous `yield*`
```

Comment:
I think we should use rules that are as similar as possible with `for` and `await for`.

Reply:
If so, then the rule should be to not invoke extension type members.

I really do not want to get into defining which part of the `Stream` interface that `await for` and `yield*` uses.

* Call `listen` with three functions arguments, and `cancelOnError: true`, get object back. (This is the simple step.)
* Occasionally call `pause()` or `pause(Future)`, maybe `resume` and `cancel` on the returned object.
* Maybe read the `isPaused` getter. (Not necessary, but can be used as optimization, to never pause more than once.)
* Possibly call `onData`, `onError`, `onDone` methods with function arguments (or maybe even with `null`). That's a valid approach to changing the continuation that the value comes back to, instead of keeping state and a trampoline on the side. 
* Maybe even call `asFuture` (but I doubt that's useful).
* Await the result of calling `cancel`.

Basically, you *have* to return something which implements the entire `StreamSubscription` interface.
For `Iterator`, that's easy - it's two members and we always call both of them.
For `StreamSubscription`, you can possibly get away with only `pause`, `resume` and `cancel`. And *if* we let you get away with that, then it becomes a breaking change for us to change the implementation in the future.

So I want to enforce that whatever `listen` returns, it implements `StreamSubscription`.

Also, the compilation of async operations is much more complicated than a simple `for`/`in` loop. The way lowering and compilation of async functions works, we are likely reusing some functionality that does class instance member invocations. Having to create new compiler paths to remember an extension type, just to call an extension type `pause` method on something which is a `StreamSubscription`, is just not worth it.

The story of "An `await for` doesn't see your static type, only the `Stream<T>` that it implements, because that's all it *needs*" is simple and consistent.
That's why that's what I went with for non-async `for`/`in` too.